openedx/core/lib/block_cache/block_structure.py

```diff
@@ -0,0 +1,588 @@
+        self._transformer_data = defaultdict(dict)
+
+    def get_xblock_field(self, usage_key, field_name, default=None):
+        """
+        Returns the collected value of the xBlock field for the
+        requested block for the requested field_name; returns default if
+        not found.
+
+        Arguments:
+            usage_key (UsageKey) - Usage key of the block whose xBlock
+                field is requested.
+
+            field_name (string) - The name of the field that is
+                requested.
+
+            default (any type) - The value to return if a field value is
+                not found.
+        """
+        block_data = self._block_data_map.get(usage_key)
+        return block_data._xblock_fields.get(field_name, default) if block_data else default
+
+    def get_transformer_data(self, transformer, key, default=None):
+        """
+        Returns the value associated with the given key from the given
+        transformer's data dictionary; returns default if not found.
+
+        Arguments:
+            transformer (BlockStructureTransformer) - The transformer
+                whose collected data is requested.
+
+            key (string) - A dictionary key to the transformer's data
+                that is requested.
+        """
+        return self._transformer_data.get(transformer.name(), {}).get(key, default)
+
+    def set_transformer_data(self, transformer, key, value):
+        """
+        Updates the given transformer's data dictionary with the given
+        key and value.
+
+        Arguments:
+            transformer (BlockStructureTransformer) - The transformer
+                whose data is to be updated.
+
+            key (string) - A dictionary key to the transformer's data.
+
+            value (any picklable type) - The value to associate with the
+                given key for the given transformer's data.
+        """
+        self._transformer_data[transformer.name()][key] = value
+
+    def get_transformer_block_data(self, usage_key, transformer, key=None, default=None):
+        """
+        Returns the value associated with the given key for the given
+        transformer for the block identified by the given usage_key;
+        returns default if not found.
+
+        Arguments:
+            usage_key (UsageKey) - Usage key of the block whose
+                transformer data is requested.
+
+            transformer (BlockStructureTransformer) - The transformer
+                whose dictionary data is requested.
+
+            key (string) - A dictionary key to the transformer's data
+                that is requested.
+
+            default (any type) - The value to return if a dictionary
+                entry is not found.
+        """
+        block_data = self._block_data_map.get(usage_key)
+        if not block_data:
+            return default
+        else:
+            transformer_data = block_data._transformer_data.get(transformer.name(), {})
```

Comment:
Since this is a `getter` method that allows passing in `default` values when the field is not present, we don't want to automatically add any data if they don't already exist.